### ISSUE TITLE
`unlock` notifies _all_ waiting tasks

### ISSUE THREAD

**Issue by kpamnany** [JuliaLang/julia#56182]
[unlock(rl::ReentrantLock)](https://github.com/JuliaLang/julia/blob/958f647d8e2d7d93b9d158f283e0891e8e9cf0b0/base/lock.jl#L187) does a [notify(cond)](https://github.com/JuliaLang/julia/blob/958f647d8e2d7d93b9d158f283e0891e8e9cf0b0/base/condition.jl#L156) which wakes all tasks waiting on the lock. As this is a lock, only one of these tasks can succeed, so it is unnecessary to [schedule](https://github.com/JuliaLang/julia/blob/958f647d8e2d7d93b9d158f283e0891e8e9cf0b0/base/condition.jl#L154) all of them. When there are a large number of tasks, the current approach wastes CPU time and can cause a quadratic amount of allocations for waitlist nodes.

It isn't straightforward to trivially fix this by changing the `notify` call to set `all=false` -- `unlock` will only call `notify` if `ReentrantLock.havelock` is `0x02` (see [here](https://github.com/JuliaLang/julia/blob/master/base/lock.jl#L197)) and that only happens after the first `notify` if all waiting tasks are woken up to race for the lock again (see [here](https://github.com/JuliaLang/julia/blob/master/base/lock.jl#L167)).

Broken out from https://github.com/JuliaLang/julia/issues/50425 as this is a separate issue.
